File: packages/ailog_flutter/ios/Classes/AilogWire.swift
import Foundation

/// Wire-format primitives shared by `Ailog` (crash-time fallback writes)
/// and `AilogFlutterPlugin`. Mirrors
/// `packages/ailog_flutter/android/.../AilogWire.kt` and, transitively,
/// the Dart algorithms in `packages/ailog/lib/src/ids.dart` and
/// `normalizer.dart`, so that a fingerprint computed here for a given
/// error type + stack frames matches the fingerprint the Dart side would
/// compute for the *same* error via `errorFingerprintFromFrames` when it
/// arrives over the MethodChannel instead. `shortHash`/
/// `errorFingerprintFromFrames`/`normalizeMessage` were verified
/// byte-for-byte against the Dart implementation (and the Kotlin port) for
/// a set of reference inputs — see `AilogWireTest.kt` on the Android side
/// for those reference values. Keep this file in sync with both if the
/// algorithm ever changes.
///
/// Note: this file could not be compiled or run in the sandbox that
/// authored it (no Xcode/macOS toolchain available). It was written to
/// mirror the Kotlin implementation, which *was* verified against Dart
/// using a real Kotlin compiler. Treat it as reviewed-but-unbuilt until a
/// real iOS toolchain compiles and exercises it.
enum AilogHash {
    private static let fnvOffsetBasis: UInt64 = 0xcbf29ce4_84222325
    private static let fnvPrime: UInt64 = 0x0000_0100_0000_01b3

    /// Non-cryptographic 64-bit FNV-1a hash, matching `fnv1a64` in Dart.
    /// Iterates UTF-16 code units to match Dart's `String.codeUnitAt`
    /// (Dart strings are UTF-16 internally).
    static func fnv1a64(_ input: String, seed: UInt64 = fnvOffsetBasis) -> UInt64 {
        var hash = seed
        for unit in input.utf16 {
            hash ^= UInt64(unit)
            hash = hash &* fnvPrime // &* wraps silently on overflow, matching Dart/Kotlin's 64-bit math.
        }
        return hash
    }

    /// Short lowercase hex token, matching `shortHash` in Dart.
    static func shortHash(_ input: String, length: Int = 8) -> String {
        let hex = String(fnv1a64(input), radix: 16)
        let padded = String(repeating: "0", count: max(0, 16 - hex.count)) + hex
        let clamped = min(max(length, 1), 16)
        return String(padded.prefix(clamped))
    }
}

private enum AilogPatterns {
    // swiftlint:disable force_try
    static let number = try! NSRegularExpression(pattern: "(?<![A-Za-z_])\\d+(\\.\\d+)?")
    static let quoted = try! NSRegularExpression(pattern: "\"[^\"]*\"|'[^']*'")
    static let url = try! NSRegularExpression(pattern: "https?://\\S+")
    static let uuid = try! NSRegularExpression(
        pattern: "\\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\\b",
        options: [.caseInsensitive]
    )
    static let path = try! NSRegularExpression(pattern: "(/[\\w.-]+){2,}")
    static let hexish = try! NSRegularExpression(pattern: "\\b[0-9a-f]{8,}\\b", options: [.caseInsensitive])
    static let whitespace = try! NSRegularExpression(pattern: "\\s+")
    // swiftlint:enable force_try
}

private func replaceAll(_ regex: NSRegularExpression, in text: String, with replacement: String) -> String {
    let range = NSRange(text.startIndex..<text.endIndex, in: text)
    return regex.stringByReplacingMatches(in: text, options: [], range: range, withTemplate: replacement)
}

/// Best-effort port of `normalizeMessage` from the Dart package. Only used
/// as the fingerprint fallback when an error has no stack frames at all
/// (rare for native exceptions); when frames are present,
/// `errorFingerprintFromFrames` never reaches this function.
func normalizeMessage(_ message: String) -> String {
    var normalized = message.lowercased()
    normalized = replaceAll(AilogPatterns.url, in: normalized, with: "<url>")
    normalized = replaceAll(AilogPatterns.uuid, in: normalized, with: "<uuid>")
    normalized = replaceAll(AilogPatterns.quoted, in: normalized, with: "<str>")
    normalized = replaceAll(AilogPatterns.path, in: normalized, with: "<path>")
    normalized = replaceAll(AilogPatterns.hexish, in: normalized, with: "<hex>")
    normalized = replaceAll(AilogPatterns.number, in: normalized, with: "<n>")
    normalized = replaceAll(AilogPatterns.whitespace, in: normalized, with: " ")
    return normalized.trimmingCharacters(in: .whitespaces)
}

/// Mirrors `errorFingerprintFromFrames` in Dart exactly for the
/// frames-present case: `type|frame1|frame2|...` hashed with `AilogHash`.
func errorFingerprintFromFrames(
    errorType: String,
    message: String,
    frames: [String],
    framesInFingerprint: Int = 5
) -> String {
    var signature = errorType
    if frames.isEmpty {
        signature += "|" + normalizeMessage(message)
    } else {
        for frame in frames.prefix(framesInFingerprint) {
            signature += "|" + frame
        }
    }
    return AilogHash.shortHash(signature)
}

/// A structured error, ready to be embedded in a JSONL `err` field.
struct NativeError {
    let type: String
    let message: String
    let frames: [String]

    init(type: String, message: String, frames: [String] = []) {
        self.type = type
        self.message = message
        self.frames = frames
    }

    var fingerprint: String {
        errorFingerprintFromFrames(errorType: type, message: message, frames: frames)
    }
}

/// Minimal recursive JSON encoder — kept dependency-free on purpose, and
/// used instead of `JSONSerialization` because the latter does not
/// guarantee key order, which matters for keeping this output readable
/// and consistent with the Dart/Kotlin writers.
///
/// Accepts `String`, `Bool`, `Int`, `Double`, `nil`, `[(String, Any?)]`
/// (ordered object) and `[Any?]` (array); anything else is stringified.
func encodeJson(_ value: Any?) -> String {
    var out = ""
    encodeJsonInto(value, &out)
    return out
}

private func encodeJsonInto(_ value: Any?, _ out: inout String) {
    switch value {
    case .none:
        out += "null"
    case let some as Bool:
        out += some ? "true" : "false"
    case let some as Int:
        out += String(some)
    case let some as Double:
        out += (some.isNaN || some.isInfinite) ? "\"\(some)\"" : String(some)
    case let some as String:
        encodeJsonString(some, &out)
    case let pairs as [(String, Any?)]:
        out += "{"
        for (index, pair) in pairs.enumerated() {
            if index > 0 { out += "," }
            encodeJsonString(pair.0, &out)
            out += ":"
            encodeJsonInto(pair.1, &out)
        }
        out += "}"
    case let array as [Any?]:
        out += "["
        for (index, item) in array.enumerated() {
            if index > 0 { out += "," }
            encodeJsonInto(item, &out)
        }
        out += "]"
    case let some?:
        encodeJsonString(String(describing: some), &out)
    }
}

private func encodeJsonString(_ value: String, _ out: inout String) {
    out += "\""
    for scalar in value.unicodeScalars {
        switch scalar {
        case "\"": out += "\\\""
        case "\\": out += "\\\\"
        case "\n": out += "\\n"
        case "\r": out += "\\r"
        case "\t": out += "\\t"
        default:
            if scalar.value < 0x20 {
                out += String(format: "\\u%04x", scalar.value)
            } else {
                out.unicodeScalars.append(scalar)
            }
        }
    }
    out += "\""
}

/// Appends `ailog`-compatible JSONL lines directly to a file, bypassing
/// the MethodChannel entirely. This exists for exactly one purpose:
/// recording an uncaught exception when the Flutter engine may already be
/// torn down and therefore cannot be relied on to relay anything. See
/// `AilogFlutterPlugin`'s uncaught exception handler, which is the only
/// normal caller.
///
/// Lines written here carry their own `ses` (generated once per process)
/// and `seq` — a distinct writer identity from the Dart `Logger`, not a
/// continuation of its session. `ailog_digest` still groups them correctly
/// because grouping is by error fingerprint, not by session.
enum AilogJsonlWriter {
    private static let sessionId: String = randomHex(bytes: 16)
    private static var sequence = 0
    private static let lock = NSLock()

    private static func isoTimestampNow() -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter.string(from: Date())
    }

    static func write(path: String?, level: String, message: String, loggerName: String, error: NativeError?) {
        guard let path = path, !path.isEmpty else { return }

        lock.lock()
        defer { lock.unlock() }

        sequence += 1
        var pairs: [(String, Any?)] = [
            ("ts", isoTimestampNow()),
            ("lvl", level),
            ("msg", message),
            ("lg", loggerName),
            ("ses", sessionId),
            ("seq", sequence),
        ]
        if let error = error {
            var errorPairs: [(String, Any?)] = [
                ("t", error.type),
                ("m", error.message),
                ("fp", error.fingerprint),
            ]
            if !error.frames.isEmpty {
                errorPairs.append(("fr", error.frames))
            }
            pairs.append(("err", errorPairs))
        }

        let line = encodeJson(pairs) + "\n"
        guard let data = line.data(using: .utf8) else { return }

        do {
            let fileURL = URL(fileURLWithPath: path)
            try FileManager.default.createDirectory(
                at: fileURL.deletingLastPathComponent(),
                withIntermediateDirectories: true
            )
        } catch {
            // Directory may already exist, or be uncreatable. Either way, fall
            // through and let open() decide — a crash-time write must never
            // throw further.
        }

        // O_APPEND, not seek-then-write. The kernel resolves the offset as
        // part of the same write() call, so a Dart isolate appending to the
        // same file concurrently can't cause this line to land at a stale
        // offset and clobber theirs. This mirrors the Kotlin writer, which
        // gets the same guarantee via FileOutputStream(file, append = true).
        let descriptor = path.withCString { cPath in
            open(cPath, O_WRONLY | O_APPEND | O_CREAT, 0o644)
        }
        guard descriptor >= 0 else { return }
        defer { close(descriptor) }

        data.withUnsafeBytes { (buffer: UnsafeRawBufferPointer) in
            guard let base = buffer.baseAddress else { return }
            var written = 0
            while written < buffer.count {
                let n = write(descriptor, base + written, buffer.count - written)
                // A short write is legal; retry the remainder. Give up on a
                // hard error rather than spinning inside a crash handler.
                if n <= 0 { break }
                written += n
            }
        }
    }
}

private func randomHex(bytes: Int) -> String {
    var result = ""
    for _ in 0..<bytes {
        result += String(format: "%02x", UInt8.random(in: 0...255))
    }
    return result
}

File: packages/ailog_flutter/android/src/main/kotlin/dev/ailog/ailog_flutter/AilogWire.kt
package dev.ailog.ailog_flutter

import java.io.File
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.atomic.AtomicInteger
import kotlin.random.Random

/**
 * Wire-format primitives shared by [Ailog] (crash-time fallback writes) and
 * [AilogFlutterPlugin]. Kept dependency-free and mirrors the algorithms in
 * ailog's Dart implementation (`packages/ailog/lib/src/ids.dart` and
 * `normalizer.dart`) so that a fingerprint computed here for a given error
 * type + stack frames matches the fingerprint the Dart side would compute
 * for the *same* error via `errorFingerprintFromFrames` when it arrives
 * over the MethodChannel instead. That parity is what lets `ailog_digest`
 * group "the same native crash" into one bucket regardless of which path
 * (channel forward vs. direct crash-time write) produced a given line.
 */
internal object AilogHash {
    private val fnvOffsetBasis = java.lang.Long.parseUnsignedLong("cbf29ce484222325", 16)
    private const val FNV_PRIME = 0x100000001b3L

    /** Non-cryptographic 64-bit FNV-1a hash, matching `fnv1a64Hex` in Dart. */
    fun fnv1a64(input: String, seed: Long = fnvOffsetBasis): Long {
        var hash = seed
        for (element in input) {
            hash = hash xor element.code.toLong()
            // Kotlin Long multiplication wraps silently. Dart computes the same
            // value via two 32-bit halves (so it also builds for web); the two
            // are verified equal against shared fixtures in AilogWireTest.
            hash *= FNV_PRIME
        }
        return hash
    }

    /** Short lowercase hex token, matching `shortHash` in Dart. */
    fun shortHash(input: String, length: Int = 8): String {
        // Long.toHexString treats the value as unsigned, so this never
        // produces a leading '-' the way naive signed formatting would.
        val hex = java.lang.Long.toHexString(fnv1a64(input)).padStart(16, '0')
        return hex.substring(0, length.coerceIn(1, 16))
    }
}

private val numberPattern = Regex("(?<![A-Za-z_])\\d+(\\.\\d+)?")
private val quotedPattern = Regex("\"[^\"]*\"|'[^']*'")
private val urlPattern = Regex("https?://\\S+")
private val uuidPattern =
    Regex("\\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\\b", RegexOption.IGNORE_CASE)
private val pathPattern = Regex("(/[\\w.-]+){2,}")
private val hexishPattern = Regex("\\b[0-9a-f]{8,}\\b", RegexOption.IGNORE_CASE)
private val whitespacePattern = Regex("\\s+")

/**
 * Best-effort port of `normalizeMessage` from the Dart package. Only used
 * as the fingerprint fallback when an error has no stack frames at all
 * (rare for native exceptions); when frames are present, [errorFingerprintFromFrames]
 * never reaches this function, so exact byte-for-byte parity with Dart's
 * ICU-backed regex engine is not required for correctness in the common
 * case.
 */
internal fun normalizeMessage(message: String): String {
    var normalized = message.lowercase(Locale.US)
    normalized = urlPattern.replace(normalized, "<url>")
    normalized = uuidPattern.replace(normalized, "<uuid>")
    normalized = quotedPattern.replace(normalized, "<str>")
    normalized = pathPattern.replace(normalized, "<path>")
    normalized = hexishPattern.replace(normalized, "<hex>")
    normalized = numberPattern.replace(normalized, "<n>")
    normalized = whitespacePattern.replace(normalized, " ").trim()
    return normalized
}

/**
 * Mirrors `errorFingerprintFromFrames` in Dart exactly for the
 * frames-present case: `type|frame1|frame2|...` hashed with [AilogHash].
 */
internal fun errorFingerprintFromFrames(
    errorType: String,
    message: String,
    frames: List<String>,
    framesInFingerprint: Int = 5,
): String {
    val signature = StringBuilder(errorType)
    if (frames.isEmpty()) {
        signature.append('|').append(normalizeMessage(message))
    } else {
        for (frame in frames.take(framesInFingerprint)) {
            signature.append('|').append(frame)
        }
    }
    return AilogHash.shortHash(signature.toString())
}

/** A structured error, ready to be embedded in a JSONL `err` field. */
internal data class NativeError(
    val type: String,
    val message: String,
    val frames: List<String> = emptyList(),
) {
    val fingerprint: String
        get() = errorFingerprintFromFrames(type, message, frames)
}

/** Minimal recursive JSON encoder — kept dependency-free on purpose. */
internal fun encodeJson(value: Any?): String {
    val builder = StringBuilder()
    encodeJsonInto(value, builder)
    return builder.toString()
}

private fun encodeJsonInto(value: Any?, out: StringBuilder) {
    when (value) {
        null -> out.append("null")
        is Boolean -> out.append(value.toString())
        is Int, is Long -> out.append(value.toString())
        is Double -> out.append(if (value.isNaN() || value.isInfinite()) "\"$value\"" else value.toString())
        is Float -> out.append(if (value.isNaN() || value.isInfinite()) "\"$value\"" else value.toString())
        is String -> encodeJsonString(value, out)
        is Map<*, *> -> {
            out.append('{')
            var first = true
            for ((key, v) in value) {
                if (!first) out.append(',')
                first = false
                encodeJsonString(key.toString(), out)
                out.append(':')
                encodeJsonInto(v, out)
            }
            out.append('}')
        }
        is Iterable<*> -> {
            out.append('[')
            var first = true
            for (item in value) {
                if (!first) out.append(',')
                first = false
                encodeJsonInto(item, out)
            }
            out.append(']')
        }
        else -> encodeJsonString(value.toString(), out)
    }
}

private fun encodeJsonString(value: String, out: StringBuilder) {
    out.append('"')
    for (c in value) {
        when (c) {
            '"' -> out.append("\\\"")
            '\\' -> out.append("\\\\")
            '\n' -> out.append("\\n")
            '\r' -> out.append("\\r")
            '\t' -> out.append("\\t")
            else -> {
                if (c.code < 0x20) {
                    out.append("\\u").append(c.code.toString(16).padStart(4, '0'))
                } else {
                    out.append(c)
                }
            }
        }
    }
    out.append('"')
}

/**
 * Appends `ailog`-compatible JSONL lines directly to a file, bypassing the
 * MethodChannel entirely. This exists for exactly one purpose: recording an
 * uncaught exception when the Flutter engine may already be torn down and
 * therefore cannot be relied on to relay anything. See
 * [AilogFlutterPlugin]'s uncaught exception handler, which is the only
 * normal caller.
 *
 * Lines written here carry their own `ses` (generated once per process) and
 * `seq` — a distinct writer identity from the Dart [Logger], not a
 * continuation of its session. `ailog_digest` still groups them correctly
 * because grouping is by error fingerprint, not by session.
 */
internal object AilogJsonlWriter {
    private val sessionId = randomHex(16)
    private val sequence = AtomicInteger(0)
    private val writeLock = Any()

    private fun isoTimestampNow(): String {
        val format =
            SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
                timeZone = TimeZone.getTimeZone("UTC")
            }
        return format.format(java.util.Date())
    }

    fun write(path: String?, level: String, message: String, loggerName: String, error: NativeError?) {
        if (path.isNullOrEmpty()) return
        val obj = LinkedHashMap<String, Any?>()
        obj["ts"] = isoTimestampNow()
        obj["lvl"] = level
        obj["msg"] = message
        obj["lg"] = loggerName
        obj["ses"] = sessionId
        obj["seq"] = sequence.incrementAndGet()
        if (error != null) {
            val errObj = LinkedHashMap<String, Any?>()
            errObj["t"] = error.type
            errObj["m"] = error.message
            errObj["fp"] = error.fingerprint
            if (error.frames.isNotEmpty()) errObj["fr"] = error.frames
            obj["err"] = errObj
        }
        val line = encodeJson(obj)

        synchronized(writeLock) {
            try {
                val file = File(path)
                file.parentFile?.mkdirs()
                file.appendText(line + "\n")
            } catch (_: Throwable) {
                // A failed crash-time write must never throw further — this
                // runs inside an uncaught exception handler.
            }
        }
    }
}

private fun randomHex(bytes: Int): String {
    val builder = StringBuilder(bytes * 2)
    repeat(bytes) {
        builder.append(Random.nextInt(0, 256).toString(16).padStart(2, '0'))
    }
    return builder.toString()
}
